gridstatus/ercot_api/ercot_api.py

```diff
@@ -1,15 +1,295 @@
+        Finds the document IDs for the emil id files that match a given date range
+
+        Args:
+            emil_id (str): the EMIL ID of the data product
+            start_date (str): the start date in the format "YYYY-MM-DD"
+            end_date (str): the end date in the format "YYYY-MM-DD"
+            verbose (bool): if True, will print out status messages
+
+        Returns:
+            list: a list of document IDs to download
+        """
+        archives = self.get_document_list_for_data_product(
+            emil_id,
+            start_date,
+            end_date,
+            size=size,
+            verbose=verbose,
+        )
+
+        if start_date is not None:
+            date_range = [str(pd.Timestamp(start_date).date())]
+            found_dates = []
+
+            if end_date:
+                date_range = [
+                    str(date.date()) for date in pd.date_range(start_date, end_date)
+                ]
+
+            document_ids = []
+
+            for doc in archives:
+                doc_posted_date = doc["postDatetime"].split("T")[0]
+                if doc_posted_date in date_range:
+                    found_dates.append(doc_posted_date)
+                    document_ids.append(doc["docId"])
+
+            log(f"Missing dates: {set(date_range) - set(found_dates)}", verbose)
+        else:
+            document_ids = [doc["docId"] for doc in archives]
+
+        return document_ids
+
+    @support_date_range(frequency=None)
+    def get_dam_lmp_hourly_by_bus(self, date, end_date=None, verbose=False):
+        """
+        Retrieves the hourly Day Ahead Market (DAM) Location Marginal Prices (LMPs)
+        for the given date range.
+
+        Data source: https://data.ercot.com/data-product-archive/NP4-183-CD
+        (requires login)
+        """
+        size = None
+
+        # Subtract one day since this is the day ahead market
+        date = date if date == "latest" else date - pd.DateOffset(days=1)
+
+        if date == "latest":
+            size = 1
+            date = None
+            end_date = None
+
+        document_ids = self.get_document_ids_to_download(
+            emil_id=DAM_LMP_HOURLY_EMIL_ID,
+            start_date=date,
+            end_date=end_date,
+            size=size,
+            verbose=verbose,
+        )
+
+        dfs = []
+
+        for doc_id in tqdm(document_ids, desc="Downloading documents"):
+            df = self.get_zip_file(
+                DAM_LMP_HOURLY_EMIL_ID,
+                document_id=doc_id,
+                verbose=verbose,
+            )
+            dfs.append(df)
+
+        data = pd.concat(dfs)
+
+        return self.parse_dam_doc(data)
+
+    def parse_dam_doc(self, data):
+        return (
+            Ercot()
+            .parse_doc(data)
+            .rename(columns={"BusName": "Location"})
+            .drop(columns=["Time"])
+            .sort_values(["Interval Start"])
+            .reset_index(drop=True)
+        )
+
 
 def hit_ercot_api(
```

Comment:
I didn't touch any of the existing code. My plan is to integrate the existing methods into the `ErcotAPI` class once we have it more solid.